[CASSANDRA-20250][trunk] Optimize Counter, Meter and Histogram metrics using thread local counters

netudima opened the pull request:
Optimize Counter, Meter and Histogram metrics using thread local counters

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20250

This is an old prototype branch.
A rebased and squashed version is available in https://github.com/apache/cassandra/pull/4324

Commit:
A very basic sketch of thread local counter metric
CASSANDRA-20250

Commit:
remove double init

Commit:
minor fixes

Commit:
fix optimistic locking logic: we should compare immutable values

Commit:
add piggyback array implementation which rely on external volatile visibility
add a basic JMH test
make ThreadLocalMetricsTest parametrized

Commit:
replace codahale Meter and Counter with ThreadLocal implementations

Commit:
integrate CASSANDRA-19332 patch for Meter, fix formatting

Commit:
fix addNonStatic method - do not do a thread local looku
add JMH for Meter implementations
switch to approximate clock to reduce time retrieval overhead

Commit:
add cache polution emulation to MetersBench
fix reset

Commit:
added more notes about possible optimizations

Commit:
remove EWMA as a separate object, move calculations to ThreadLocalExponentialMovingAverages, use single uncounted value for all 3 rates

Commit:
add unit tests to check Meter behavior using codahale Meter as a standard

Commit:
replace codahale Timer with ThreadLocalTimer
move tickIfNessesary out from Meter.mark logic to a periodic background task
remove LazySetArrayThreadLocalMetrics implementation, rename PiggybackArrayThreadLocalMetrics to ThreadLocalMetrics

Commit:
replace codahale Histogram with ThreadLocalHistogram
use atomic long for hints in-procress counter which is frequently read
introduce common Destroyable interface for metrics
move cleanDeadAndUpdateSummaries invocations to a background thread and metric creation/destoying logic

Commit:
use shared non-thread local double[] for rates to improve memory locality
move ThreadLocalExponentialMovingAverages logic to ThreadLocalMeter itself

Commit:
Meters may have different tick values if they created at different moments, so we have to keep tick values per instance
Fix condition in allocateRateGroupOffset
Add dynamic Meter creation/destroying emulation

Commit:
use PhantomReferences to recycle metric IDs and ThreadLocalMetrics objects associated with dead threads
use BitSet for free metric id set

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20251

Commit:
add javadoc for ThreadLocalMetrics, fix some concurrency issues related to metrics and threads lifecycles

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20251

Commit:
add javadoc for ThreadLocalMeter, use BitSet for free rate group ids

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20251

Commit:
add more javadocs, improve test coverage, move LongAdderCounter to tests

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20251

Commit:
few more asserts

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20251

clohfink reviewed (commented):
Wouldn’t using longadder instead of atomic longs do the same thing without having to create a layer over everything?

netudima commented:
unfortunately no, it is already LongAdder used in Dropwizard but it is not for free from CPU usage for our amount of metrics + it has a high memory overhead: [CASSANDRA-20250](https://issues.apache.org/jira/browse/CASSANDRA-20250) - here there is a long conversation with Benedict about it and different perf test results.

netudima commented:
A similar idea is also described by Nitsan: https://psy-lob-saw.blogspot.com/2013/06/java-concurrent-counters-by-numbers.html (as ThreadLocalCounter option)

Commit:
Inherit ThreadLocal metric classes from Dropwizard metrics to provide backward compatibility for Dropwizard MetricRegistryListener and Reporter functionality.

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20250

Commit:
Exclude "ThreadLocalMetrics-Cleaner" thread from unit test checks

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20250

Commit:
Inherit AtomicLongCounter from Dropwizard Counter as well
Fix concurrency issues in writing from multiple threads to ThreadLocalMeter.rates

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20250

Commit:
use different name for AtomicLongCounter field to be able to null parent class field separately

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20250

netudima commented:
This is an old prototype branch.
A rebased and squashed version is available in https://github.com/apache/cassandra/pull/4324
